include/all.h

```diff
@@ -7,3 +9,8 @@
 #include <unique.hpp>
 #endif
 #endif
+
+#define STRINGIFY(x) #x
+#define TOSTRING(x) STRINGIFY(x)
+#define SOURCE_LOC " (" __FILE__ ":" TOSTRING(__LINE__) ")"
```

Comment:
It occurs to me that this solution is also useful with build/verbosity macros. Should use those here to conditionally define `SOURCE_LOC` only for debug builds. In production, this will produce paths that probably aren't meaningful, or at least would only be useful in a bug report with a version we could check out and find the right relative file from the absolute paths this reports.

Reply:
I agree, just thought it might be odd if you have a build system creating binaries that then report paths that don't even exist when an error is logged.  I'm good with it as is for the time being.

Reply:
Yeah, that's not uncommon in my experience--potentially a little confusing but you quickly figure out what's going on. Something that just occurs to me... should we try [embedding the Git commit ID into the file](https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes#_keyword_expansion) and hence the message?

Reply:
Eh...that's messier than it at first looks. Maybe not, or maybe later.